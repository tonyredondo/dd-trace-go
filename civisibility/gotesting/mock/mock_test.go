// Unless explicitly stated otherwise all files in this repository are licensed
// under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2024 Datadog, Inc.

package mock

import (
	"os"
	"testing"
	"time"
	_ "unsafe"

	"github.com/gkampitakis/go-snaps/match"
	"github.com/gkampitakis/go-snaps/snaps"
	"gopkg.in/DataDog/dd-trace-go.v1/civisibility/gotesting"
	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/mocktracer"
	ddtracer "gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
	"gopkg.in/DataDog/dd-trace-go.v1/internal/civisibility"
)

var tracer mocktracer.Tracer

type SSpan struct {
	Name       string
	Tags       map[string]any
	FinishTime time.Time
	StartTime  time.Time
	SpanId     uint64
	TraceId    uint64
	ParentId   uint64
}

func TestMain(om *testing.M) {
	// Initialize civisibility using the mocktracer for testing
	tracer = civisibility.InitializeCiVisibilityMock()
	m := (*gotesting.M)(om)
	os.Exit(m.Run())
}

func TestDummyTestForMocking(ot *testing.T) {
	t := (*gotesting.T)(ot)
	t.Run("Child", func(ot *testing.T) {
		t := (*gotesting.T)(ot)
		if span, ok := ddtracer.SpanFromContext(t.Context()); ok {
			span.SetTag("Custom Tag", "Custom Value")
		}

		span, _ := ddtracer.StartSpanFromContext(t.Context(), "Custom Span")
		defer span.Finish()
		span.SetTag("Key", "Value")
	})
}

func TestAssertMock(t *testing.T) {
	spans := tracer.FinishedSpans()
	if len(spans) == 0 {
		t.Error("No mock spans found")
	}

	serializableSpans := make([]SSpan, len(spans))
	for i, span := range spans {
		serializableSpans[i] = SSpan{
			Name:       span.OperationName(),
			Tags:       span.Tags(),
			FinishTime: span.FinishTime(),
			StartTime:  span.StartTime(),
			SpanId:     span.SpanID(),
			TraceId:    span.TraceID(),
			ParentId:   span.ParentID(),
		}
	}

	snaps.MatchJSON(t, serializableSpans,
		match.Any(
			"#.StartTime",
			"#.FinishTime",
			"#.Tags.ci\\.workspace_path",
			"#.Tags.git\\.branch",
			"#.Tags.git\\.commit\\.author\\.date",
			"#.Tags.git\\.commit\\.author\\.email",
			"#.Tags.git\\.commit\\.author\\.name",
			"#.Tags.git\\.commit\\.committer\\.date",
			"#.Tags.git\\.commit\\.committer\\.email",
			"#.Tags.git\\.commit\\.committer\\.name",
			"#.Tags.git\\.commit\\.message",
			"#.Tags.git\\.commit\\.sha",
			"#.Tags.git\\.repository_url",
			"#.Tags.os\\.architecture",
			"#.Tags.os\\.platform",
			"#.Tags.os\\.version",
			"#.Tags.runtime\\.version",
			"#.Tags.test\\.command",
			"#.Tags.test\\.framework_version",
			"#.Tags.test\\.source\\.start"))
}
