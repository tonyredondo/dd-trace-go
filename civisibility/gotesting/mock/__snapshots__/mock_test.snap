
[TestAssertMock - 1]
[
 {
  "FinishTime": "<Any value>",
  "Name": "Custom Span",
  "ParentId": 128,
  "SpanId": 129,
  "StartTime": "<Any value>",
  "Tags": {
   "Key": "Value",
   "resource.name": "Custom Span",
   "service.name": null
  },
  "TraceId": 128
 },
 {
  "FinishTime": "<Any value>",
  "Name": "golang.org/pkg/testing.test",
  "ParentId": 0,
  "SpanId": 128,
  "StartTime": "<Any value>",
  "Tags": {
   "Custom Tag": "Custom Value",
   "_dd.origin": "ciapp-test",
   "ci.workspace_path": "<Any value>",
   "git.branch": "<Any value>",
   "git.commit.author.date": "<Any value>",
   "git.commit.author.email": "<Any value>",
   "git.commit.author.name": "<Any value>",
   "git.commit.committer.date": "<Any value>",
   "git.commit.committer.email": "<Any value>",
   "git.commit.committer.name": "<Any value>",
   "git.commit.message": "<Any value>",
   "git.commit.sha": "<Any value>",
   "git.repository_url": "<Any value>",
   "manual.keep": true,
   "os.architecture": "<Any value>",
   "os.platform": "<Any value>",
   "os.version": "<Any value>",
   "resource.name": "mock_test.go.TestDummyTestForMocking/Child",
   "runtime.name": "gc",
   "runtime.version": "<Any value>",
   "span.type": "test",
   "test.codeowners": "[\"@DataDog/dd-trace-go-guild\"]",
   "test.command": "<Any value>",
   "test.framework": "golang.org/pkg/testing",
   "test.framework_version": "<Any value>",
   "test.module": "gopkg.in/DataDog/dd-trace-go.v1/civisibility/gotesting/mock",
   "test.name": "TestDummyTestForMocking/Child",
   "test.source.file": "civisibility/gotesting/mock/mock_test.go",
   "test.source.start": "<Any value>",
   "test.status": "pass",
   "test.suite": "mock_test.go",
   "test.type": "test",
   "test.working_directory": "civisibility/gotesting/mock",
   "test_module_id": "125",
   "test_session_id": "124",
   "test_suite_id": "126"
  },
  "TraceId": 128
 },
 {
  "FinishTime": "<Any value>",
  "Name": "golang.org/pkg/testing.test",
  "ParentId": 0,
  "SpanId": 127,
  "StartTime": "<Any value>",
  "Tags": {
   "_dd.origin": "ciapp-test",
   "ci.workspace_path": "<Any value>",
   "git.branch": "<Any value>",
   "git.commit.author.date": "<Any value>",
   "git.commit.author.email": "<Any value>",
   "git.commit.author.name": "<Any value>",
   "git.commit.committer.date": "<Any value>",
   "git.commit.committer.email": "<Any value>",
   "git.commit.committer.name": "<Any value>",
   "git.commit.message": "<Any value>",
   "git.commit.sha": "<Any value>",
   "git.repository_url": "<Any value>",
   "manual.keep": true,
   "os.architecture": "<Any value>",
   "os.platform": "<Any value>",
   "os.version": "<Any value>",
   "resource.name": "mock_test.go.TestDummyTestForMocking",
   "runtime.name": "gc",
   "runtime.version": "<Any value>",
   "span.type": "test",
   "test.codeowners": "[\"@DataDog/dd-trace-go-guild\"]",
   "test.command": "<Any value>",
   "test.framework": "golang.org/pkg/testing",
   "test.framework_version": "<Any value>",
   "test.module": "gopkg.in/DataDog/dd-trace-go.v1/civisibility/gotesting/mock",
   "test.name": "TestDummyTestForMocking",
   "test.source.file": "civisibility/gotesting/mock/mock_test.go",
   "test.source.start": "<Any value>",
   "test.status": "pass",
   "test.suite": "mock_test.go",
   "test.type": "test",
   "test.working_directory": "civisibility/gotesting/mock",
   "test_module_id": "125",
   "test_session_id": "124",
   "test_suite_id": "126"
  },
  "TraceId": 127
 }
]
---
